### PR TITLE
COL-50 - Implement commenting REST services

### PR DESCRIPTION
**Pull request by nicolaasmatthijs**
- List comments for an asset
- Create comment REST service
- Reply to a comment

https://jira.ets.berkeley.edu/jira/browse/COL-50

### CODE REVIEW

**Comment by nicolaasmatthijs** [comment]
Added editing a comment

**Comment by nicolaasmatthijs** [comment]
Rebased against master

File: node_modules/col-assets/lib/api.js [hunk 1]
```diff
@@ -187,3 +201,175 @@ var createLink = module.exports.createLink = function(ctx, title, url, opts, cal
...
+      'id': id
+    },
+    'include': [{
+      'model': DB.User,
+      'attributes': ['id', 'canvas_user_id', 'canvas_course_role', 'canvas_full_name', 'canvas_image']
+    }]
+  };
+  DB.Comment.find(options).complete(function(err, comment) {
+    if (err) {
+      log.error({'err': err}, 'Failed to retrieve the created comment');
```
Comment (simong) [inline node_modules/col-assets/lib/api.js]: Maybe log the `id` here as well

File: node_modules/col-assets/lib/api.js [hunk 2]
```diff
@@ -187,3 +201,175 @@ var createLink = module.exports.createLink = function(ctx, title, url, opts, cal
...
+
+    // Check that the comment to which this comment is a reply exists
+    if (parent) {
+      // Ensure the parent id is a number
+      parent = CollabosphereUtil.getNumberParam(parent);
+      var parentComment = _.findWhere(asset.toJSON().comments, {'id': parent});
+      if (!parentComment) {
+        log.error({'err': err, 'parent': parent}, 'Failed to find the comment to which a reply is being made');
+        return callback({'code': 400, 'msg': 'Failed to find the comment to which a reply is being made'});
+      }
```
Comment (simong) [inline node_modules/col-assets/lib/api.js]: Should we do a check to see if the `parentComment` has no `parent` of its own? IIRC we wouldn't allow for further nesting right?

File: node_modules/col-assets/lib/rest.js [hunk 3]
```diff
@@ -70,3 +70,29 @@ Collabosphere.apiRouter.post('/assets', function(req, res) {
...
+ */
+Collabosphere.apiRouter.post('/assets/:assetId/comments/:commentId', function(req, res) {
+  AssetsAPI.editComment(req.ctx, req.params.assetId, req.params.commentId, req.body.body, function(err, comment) {
+    if (err) {
+      return res.status(err.code).send(err.msg);
+    }
+
+    return res.status(200).send(comment);
+  });
+});
```
Comment (simong) [inline node_modules/col-assets/lib/rest.js]: Should students be able to delete comments?

**Comment by simong** [comment]
Couple of small comments. Back to @nicolaasmatthijs

File: node_modules/col-assets/lib/api.js (hunk 2 above)
Comment (nicolaasmatthijs) [inline node_modules/col-assets/lib/api.js]: We'll allow it through the REST APIs. The UI just won't expose it. I even have a test that checks that it works.

File: node_modules/col-assets/lib/rest.js (hunk 3 above)
Comment (nicolaasmatthijs) [inline node_modules/col-assets/lib/rest.js]: Deleting comments is currently to be determined, which is why I left it out of this PR.

**Comment by nicolaasmatthijs** [comment]
Followed up on code review (including `CSRF` support)

**Comment by simong** [comment]
@nicolaasmatthijs is this ready for another round of review?

**Comment by nicolaasmatthijs** [comment]
Yep, all yours :)

**Comment by nicolaasmatthijs** [comment]
Added tests for the CSRF validation. Back to @simong for review and merge.

**Comment by nicolaasmatthijs** [comment]
@paulkerschen : I was thinking it might make sense for you to perform the last bit of review here. The PR has been mostly vetted, it was only missing some test coverage around the CSRF protection. This remaining bit of code is a nice bite-size piece of code and should give you some exposure to the test framework.

**Comment by paulkerschen** [comment]
@nicolaasmatthijs : Thanks, I'll give it a go.

**Pull request merged**